proposals/NNNN-custom-isolation-checking-for-serialexecutor.md

```diff
@@ -0,0 +1,182 @@
+
+### Extended executor comparison mechanism
+
+With this proposal, the logic for checking if the "current" executor is the same as the "expected" executor changes becomes as follows:
+
+- obtain current executor
+  - if no current executor exists, ​​use heurystics to detect the "main actor" executor 
+    - These heurystics could be removed by using this proposal's `checkIsolation()` API, however we'll first need to expose the MainActor's SerialExecutor as a global property which this proposal does not cover. Please see **Future Directions** for more discussion of this topic.
+  - if a current executor exists, perform basic object comparison between them
+- if unable to prove the executors are equal:
+  - compare the executors using "complex equality" (see [SE-0392: Custom Actor Executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md) for a detailed description of complex exector equality)
+- if still unable to prove the executors are equal:
+  - :arrow_right: call the expected executor's `checkIsolation()` method
+
+The last step of this used to be just to unconditionally fail the comparison, leaving no space for an executor to take over and use whatever it's own tracking -- usually expressed using thread-locals the executor sets as it creates its own worker thread -- to actually save the comparison from failing.
+
+Specific use-cases of this API include `DispatchSerialQueue`, which would be able to implement the requirement as follows:
+
+```swift
+// Dispatch 
+
+extension DispatchSerialQueue { 
+  public func checkIsolated(message: String) {
+    dispatchPrecondition(condition: .onQueue(self)) // existing Dispatch API
+  }
+}
+```
+
+Other executors would have the same capability, if they used some mechanisms to identify their own worker threads.
+
+### Impact on async code and isolation assumtions
+
+The `assumeIsolated(_:file:line:)` APIs purposefully only accept a **synchronous** closure. This is correct, and with the here proposed additions, it remains correct -- we may be executing NOT inside a Task, however we may be isolated and can safely access actor state. 
+
+This means that the following code snippet, while a bit unusual remains correct isolation-wise:
+
+```swift
+actor Worker {
+  var number: Int 
+  
+  nonisolated func canOnlyCallMeWhileIsolatedOnThisInstance() -> Int {
+    self.preconditionIsolated("This method must be called while isolated to \(self)")
+
+    return self.assumeIsolated { // () throws -> Int
+      // suspensions are not allowed in this closure.
+      
+      self.number // we are guaranteed to be isolated on this actor; read is safe
+    }
+  }
+      
+```
+
+As such, there is no negative impact on the correctness of these APIs.
```

Comment:
```suggestion
As such, there is no negative impact on the correctness of these APIs.

Asynchronous functions should not use isolation checking.  Isolation
checking is useful in synchronous functions because they naturally inherit
execution properties like their caller's isolation without disturbing it.
A synchronous function may be formally non-isolated and yet actually
run in an isolated context dynamically.  This is not true for asynchronous
functions, which switch to their formal isolation on entry without regard
to their caller's isolation.  If an asynchronous function is not formally
isolated to an actor, its execution will never be dynamically in an
isolated context, so there's no point in checking for it.
```

Reply:
```suggestion
As such, there is no negative impact on the correctness of these APIs.

Asynchronous functions should not use dyanamic isolation checking.  Isolation
checking is useful in synchronous functions because they naturally inherit
execution properties like their caller's isolation without disturbing it.
A synchronous function may be formally non-isolated and yet actually
run in an isolated context dynamically.  This is not true for asynchronous
functions, which switch to their formal isolation on entry without regard
to their caller's isolation.  If an asynchronous function is not formally
isolated to an actor, its execution will never be dynamically in an
isolated context, so there's no point in checking for it.
```

added "dynamic", happy to take this edit otherwise